File: packages/web/src/api/sop.ts
import api from "./index";

export interface SOPTemplate {
  id: number;
  name: string;
  version: string;
  steps: Record<string, unknown>[];
  product_model: string | null;
  description: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export const sopApi = {
  list: () => api.get<SOPTemplate[]>("/sop/"),
  get: (id: number) => api.get<SOPTemplate>(`/sop/${id}`),
  create: (data: Partial<SOPTemplate>) => api.post<SOPTemplate>("/sop/", data),
  update: (id: number, data: Partial<SOPTemplate>) => api.put<SOPTemplate>(`/sop/${id}`, data),
  delete: (id: number) => api.delete(`/sop/${id}`),
};

File: packages/server/src/schemas/sop.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class SOPCreate(BaseModel):
    name: str = Field(..., max_length=100)
    version: str = "1.0"
    steps: list[dict[str, Any]]
    product_model: str | None = None
    description: str = ""


class SOPResponse(BaseModel):
    id: int
    name: str
    version: str
    steps: list[dict[str, Any]]
    product_model: str | None
    description: str
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
